core/src/raw/oio/write/position_write.rs

```diff
@@ -0,0 +1,283 @@
+impl<W: PositionWrite> PositionWriter<W> {
+    /// Create a new PositionWriter.
+    pub fn new(inner: W, executor: Option<Executor>, concurrent: usize) -> Self {
+        let executor = executor.unwrap_or_default();
+
+        Self {
+            w: Arc::new(inner),
+            executor: executor.clone(),
+            next_offset: 0,
+            cache: None,
+
+            tasks: ConcurrentTasks::new(executor, concurrent, |input| {
+                Box::pin(async move {
+                    let fut = input.w.write_all_at(input.offset, input.bytes.clone());
+                    match input.executor.timeout() {
+                        None => {
+                            let result = fut.await;
+                            (input, result)
+                        }
+                        Some(timeout) => {
+                            let result = select! {
+                                result = fut.fuse() => {
+                                    result
+                                }
+                                _ = timeout.fuse() => {
+                                      Err(Error::new(
+                                            ErrorKind::Unexpected, "write position timeout")
+                                                .with_context("offset", input.offset.to_string())
+                                                .set_temporary())
+                                }
+                            };
+                            (input, result)
+                        }
+                    }
+                })
+            }),
+        }
+    }
+
+    fn fill_cache(&mut self, bs: Buffer) -> usize {
+        let size = bs.len();
+        assert!(self.cache.is_none());
+        self.cache = Some(bs);
+        size
+    }
+}
+
+impl<W: PositionWrite> oio::Write for PositionWriter<W> {
+    async fn write(&mut self, bs: Buffer) -> Result<usize> {
+        if self.cache.is_none() {
+            let size = self.fill_cache(bs);
+            return Ok(size);
+        }
+
+        let bytes = self.cache.clone().expect("pending write must exist");
+        let length = bytes.len() as u64;
+        let offset = self.next_offset;
+
+        self.tasks
+            .execute(WriteInput {
+                w: self.w.clone(),
+                executor: self.executor.clone(),
+                offset,
+                bytes,
+            })
+            .await?;
+        self.cache = None;
+        self.next_offset += length;
+        let size = self.fill_cache(bs);
+        Ok(size)
+    }
+
+    async fn close(&mut self) -> Result<()> {
+        // Make sure all tasks are finished.
+        while self.tasks.next().await.transpose()?.is_some() {}
+
+        if let Some(buffer) = self.cache.clone() {
+            let offset = self.next_offset;
+            self.w.write_all_at(offset, buffer).await?;
+            self.cache = None;
+        }
+        Ok(())
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        Ok(())
```

Comment:
We might consider adding a `truncate` API to discard existing written data. Let's table this feature for future discussion. Would you like to create an issue for this?

Reply:
I created a new issue for this.

Reply:
@Xuanwo Have any problem yet?